crypto/fipsmodule/hmac/hmac.c

```diff
@@ -120,119 +175,114 @@ void HMAC_CTX_free(HMAC_CTX *ctx) {
+      pad[i] = 0x3636363636363636 ^ key_block[i];
     }
-
-    for (size_t i = 0; i < EVP_MAX_MD_BLOCK_SIZE; i++) {
-      pad[i] = 0x36 ^ key_block[i];
-    }
-    if (!EVP_DigestInit_ex(&ctx->i_ctx, md, impl) ||
-        !EVP_DigestUpdate(&ctx->i_ctx, pad, EVP_MD_block_size(md))) {
+    if (!methods->init(&ctx->i_ctx) ||
+        !methods->update(&ctx->i_ctx, pad, block_size)) {
       goto end;
     }
-
-    for (size_t i = 0; i < EVP_MAX_MD_BLOCK_SIZE; i++) {
-      pad[i] = 0x5c ^ key_block[i];
+    for (size_t i = 0; i < block_size / 8; i++) {
+      pad[i] = 0x5c5c5c5c5c5c5c5c ^ key_block[i];
     }
-    if (!EVP_DigestInit_ex(&ctx->o_ctx, md, impl) ||
-        !EVP_DigestUpdate(&ctx->o_ctx, pad, EVP_MD_block_size(md))) {
+    if (!methods->init(&ctx->o_ctx) ||
+        !methods->update(&ctx->o_ctx, pad, block_size)) {
       goto end;
     }
-
-    ctx->md = md;
-  }
-
-  if (!EVP_MD_CTX_copy_ex(&ctx->md_ctx, &ctx->i_ctx)) {
-    goto end;
+    ctx->initialized = 1;
   }
-  ret = 1;
-
+  OPENSSL_memcpy(&ctx->md_ctx, &ctx->i_ctx, sizeof(ctx->i_ctx));
+  result = 1;
 end:
   FIPS_service_indicator_unlock_state();
-  return ret;
+  return result;
 }
 
 int HMAC_Update(HMAC_CTX *ctx, const uint8_t *data, size_t data_len) {
-  return EVP_DigestUpdate(&ctx->md_ctx, data, data_len);
+  const HmacMethods *methods = ctx->methods;
+  if (methods == NULL || ctx->initialized != 1) {
+    return 0;
+  }
+  return methods->update(&ctx->md_ctx, data, data_len);
 }
 
 int HMAC_Final(HMAC_CTX *ctx, uint8_t *out, unsigned int *out_len) {
-  // We have to avoid the underlying SHA services updating the indicator state,
-  // so we lock the state here.
-  FIPS_service_indicator_lock_state();
-  unsigned int i;
-  uint8_t buf[EVP_MAX_MD_SIZE];
-
-  // TODO(davidben): The only thing that can officially fail here is
-  // |EVP_MD_CTX_copy_ex|, but even that should be impossible in this case.
-  if (!EVP_DigestFinal_ex(&ctx->md_ctx, buf, &i) ||
-      !EVP_MD_CTX_copy_ex(&ctx->md_ctx, &ctx->o_ctx) ||
-      !EVP_DigestUpdate(&ctx->md_ctx, buf, i) ||
-      !EVP_DigestFinal_ex(&ctx->md_ctx, out, out_len)) {
-    *out_len = 0;
-    FIPS_service_indicator_unlock_state();
+  const HmacMethods *methods = ctx->methods;
+  if (methods == NULL || ctx->initialized != 1) {
     return 0;
   }
-
+  FIPS_service_indicator_lock_state();
+  int result = 0;
+  const EVP_MD *evp_md = ctx->md;
+  int hmac_len = EVP_MD_size(evp_md);
+  uint8_t tmp[EVP_MAX_MD_SIZE];
+  if (!methods->digest(tmp, &ctx->md_ctx)) {
+    goto end;
+  }
+  OPENSSL_memcpy(&ctx->md_ctx, &ctx->o_ctx, sizeof(ctx->o_ctx));
+  if (!ctx->methods->update(&ctx->md_ctx, tmp, hmac_len)) {
+    goto end;
+  }
+  result = methods->digest(out, &ctx->md_ctx);
+end:
   FIPS_service_indicator_unlock_state();
-  HMAC_verify_service_indicator(ctx->md);
-  return 1;
+  if (result) {
+    HMAC_verify_service_indicator(evp_md);
+    if (out_len) {
+      *out_len = hmac_len;
+    }
+    return 1;
+  } else {
+    return 0;
```

Comment:
I think here `out_len` should be set to 0 to match the original behaviour.

Reply:
Will restore.